src/Api/UriVariablesConverter.php

```diff
@@ -71,7 +73,7 @@ private function getIdentifierTypes(string $resourceClass, array $properties): a
         foreach ($properties as $property) {
             $propertyMetadata = $this->propertyMetadataFactory->create($resourceClass, $property);
             foreach ($propertyMetadata->getBuiltinTypes() as $type) {
-                $types[] = Type::BUILTIN_TYPE_OBJECT === ($builtinType = $type->getBuiltinType()) ? $type->getClassName() : $builtinType;
+                $types[$property][] = Type::BUILTIN_TYPE_OBJECT === ($builtinType = $type->getBuiltinType()) ? $type->getClassName() : $builtinType;
```

Comment:
I found the issue, this PR makes an assumption that the parameterName is always in sync with properties, which is not the case as shown here https://github.com/api-platform/core/blob/main/tests/Fixtures/TestBundle/Resources/config/api_resources_odm/resources.xml#L14